Machine abstraction

lrvick opened the pull request:
## Changed
* refactors container to install fixed versions of apt packages to be more deterministic
* Adds docker-machine backend allowing one to select "local" "digitalocean" or "virtualbox" backends with ability to easily add more in the future.
* Include cloud management tooling for future cloud backends and setup pattern to hash lock them
* Refactor paths to deal with different mounting needs of remote vs local so that:
  * all configuration ends up in a single "config" tree downstreams can override
  * all downloaded sources and build cache lives in top level 'build' directory
  * all output goes to top level release directory

## TODO
- [ ] test digitalocean build support
- [x] document new backends and how to select them
- [ ] automatically download/cache docker-machine binary with wrapper script

lrvick committed:
initial refactor

lrvick committed:
Merge branch '10.0' into machine-abstraction

lrvick committed:
overhaul/refactor makefile for multiple backends (WIP)

lrvick committed:
finish refactoring allowing remote/local distinctions and fix paths

lrvick committed:
virtualbox backend support

lrvick committed:
another pile of refactoring

lrvick committed:
typos and chown fix

lrvick committed:
fix volume wrapper

lrvick committed:
overhaul container to be more deterministic + cloud management tools for future

lrvick committed:
add rough initial digitalocean support

lrvick committed:
set permissions before writing to build

lrvick committed:
Improve build caching in docker container

drGrove commented on Makefile:
```diff
@@ -1,107 +1,252 @@
-.DEFAULT_GOAL := default
-.PHONY: default
-default: build
+CPUS := $(shell nproc)
```
nproc doesn't exist on Mac OS. Since this is in Docker/VirtualBox everything _should_ have worked normally. I'd recommend adding a check for Darwin to keep the API working.

lrvick replied:
Sugesstions for an alternative command on osx? we chould just "nproc | some-osx-bs"

drGrove commented on Makefile:
```diff
@@ -1,107 +1,252 @@
+		--driver $(BACKEND) $(NAME)
+
+.PHONY: machine-install
+machine-install:
```
Drop this if you're not using it

drGrove commented on scripts/environment:
```diff
@@ -12,9 +12,9 @@ libfaketime = "/usr/lib/x86_64-linux-gnu/faketime/libfaketime.so.1"
     base_dir = environ['BASE_DIR']
 else:
-    base_dir = "%s/base" % cwd
+    base_dir = "%s/build/base" % cwd
```
Use format() and `{}`

drGrove commented on scripts/environment:
```diff
@@ -12,9 +12,9 @@ libfaketime = "/usr/lib/x86_64-linux-gnu/faketime/libfaketime.so.1"
+    base_dir = "%s/build/base" % cwd
 
-config_file='%s/config.yml' % cwd
+config_file='%s/config/config.yml' % cwd
```
use format() and `{}`

drGrove commented on scripts/environment:
```diff
@@ -70,7 +70,7 @@ if 'DEVICE' in environ and len(environ['DEVICE']) > 0:
         environment[key.upper()]=value.lower()
     if variant == "user" or variant == "userdebug":
-        key_dir = "%s/keys/%s" % (cwd, device)
+        key_dir = "%s/config/keys/%s" % (cwd, device)
```
let's upgrade this to python3 while we're already making lots of changes

lrvick committed:
add config syncing via sftp, volume fixes, fix fetch regressions

lrvick committed:
more DO cleanup, volume remove on mrproper

lrvick committed:
remove now pointless rexray conf files

dan-v committed:
make kernel build optional, make platform_ref device specific in config and add mechanism to auto update

lrvick committed:
several fixes for from-scratch builds

lrvick committed:
bump to 10.0.0_r3

lrvick committed:
bump configs for our own fork of android-prepare-vendor

lrvick committed:
refactor for new OTA+factory vendor setup

lrvick committed:
fix build vars

lrvick committed:
re-fix vars

lrvick committed:
bump to QP1A.191005.007

lrvick committed:
overhaul key generation and entropy paths for apex keys

dan-v committed:
replace subprocess with gitpython for ls_remote call to find latest tags

lrvick committed:
Merge branch 'updates' of https://github.com/RattlesnakeOS/aosp-build into RattlesnakeOS-updates

lrvick committed:
Merge branch 'RattlesnakeOS-updates' into machine-abstraction

lrvick committed:
update keys script for apex keys

ypid committed:
Sort GIT_REFs

ypid committed:
Ensure Docker image build fails early

There is a GIT_REF mismatch currently. But this only hard fails the
image build after kubernets and so on has been cloned.

ypid commented on config/container/Dockerfile:
```diff
@@ -0,0 +1,91 @@
+ARG HELM_GIT_REF="e70f012bb6b7b4aa19aa7190d68e182cb1f5b9cb"
+ARG TERRAFORM_GIT_REF="4ebf9082cde695a4bbc908cbf373a2c8139fe534"
+ARG TERRAFORM_HELM_GIT_REF="a8e657341b68c09f901736974aee3aeccdd38be9"
+ARG TERRAFORM_KUBERNETES_GIT_REF="56bb8e012f5f4bdd2fb906e1e6a567e3e57b46a4"
```
What is the reason for listing the GIT_REFs as ARG here instead of just writing it directly in the `printf` command? My guess is that you want to allow overwriting it using the `--build-arg` option? Is that actually used/needed?

drGrove replied:
It's a nice abstraction that allows for testing upgrades in the future as well as something we do in other projects

lrvick replied:
Yeah it allows you to override these in a separate file via --build-arg.

The intention is to have "make config" be able to bump these automatically in a future revision

ypid committed:
Fix GIT_REF mismatch. Did not cause in image build error because unused.

ypid committed:
Fix typo

ypid committed:
Have two empty lines between build stages to visually separate them

ypid committed:
Support to overwrite or remove the image digest, needed for local builds

Example:

```Shell
make IMAGE_OPTIONS='--build-arg UBUNTU_IMAGE_REF=' image
```

Another way would be to build a local image ubuntu:cosmic which includes
a digest and specify that, but there is no easy way to do this:
https://github.com/moby/moby/issues/32016

ypid committed:
Support to build minimal Docker image without go stuff

ypid committed:
[scripts/manifest] Retry on "fatal: unable to access" errors

I was hitting them constantly and thus was unable to have a full run of the script. It always terminated for one of the hundreds of repos. This retry fixed it.

```
Traceback (most recent call last):
  File "/home/build/scripts/manifest", line 196, in <module>
    remove_groups = config.get('platform',{}).get('remove_groups',[]),
  File "/home/build/scripts/manifest", line 40, in __init__
    self._lock()
  File "/home/build/scripts/manifest", line 145, in _lock
    remote_refs = self._lsremote(project_repo_url)
  File "/home/build/scripts/manifest", line 152, in _lsremote
    for ref in g.ls_remote(url).split('\n'):
  File "/usr/lib/python3/dist-packages/git/cmd.py", line 548, in <lambda>
    return lambda *args, **kwargs: self._call_process(name, *args, **kwargs)
  File "/usr/lib/python3/dist-packages/git/cmd.py", line 1014, in _call_process
    return self.execute(call, **exec_kwargs)
  File "/usr/lib/python3/dist-packages/git/cmd.py", line 825, in execute
    raise GitCommandError(command, status, stderr_value, stdout_value)
git.exc.GitCommandError: Cmd('git') failed due to: exit code(128)
  cmdline: git ls-remote https://android.googlesource.com/platform/external/antlr.git
  stderr: 'fatal: unable to access 'https://android.googlesource.com/platform/external/antlr.git/': The requested URL returned error: 500'
```

ypid commented on scripts/environment:
```diff
@@ -23,14 +23,14 @@ patches = ", ".join(config.get('platform',{}).get('patches',[]))
-    "PATCH_DIR": "%s/patches" % cwd,
-    "EXTERNAL_DIR": "%s/external" % cwd,
+    "MANIFEST_REPO": "%s/config/manifests" % cwd,
+    "MANIFEST_DIR": "%s/config/manifests" % cwd,
```
Seems one of those can be removed now that they are the same?

ypid committed:
[scripts/config] Fix html parsing for sailfish

Ref: 60c6f7abbdc64d038477be47ccde643cfb81c7d1

ypid committed:
[scripts/config] Increase number of tags to look back because of Pixel 4

ypid committed:
[scripts/manifest] Fix Bash/Python difference bug related to kernel xml

ypid reviewed (commented):
Impressive work @lrvick :+1:

I like the idea to pin the Docker images and also the apt packages. Also, I like the overall structure of build system so that https://github.com/hashbang/os or other ROMs can use one robust build system.

I left a few small comments. Additionally, I interactively reviewed this PR, fixed a few things and made it work for me. For example, I don’t want to have some big go projects in my TCB so I worked out a way that should be suitable for us all. For people who want to integrate this with Kubernets and so on. And for me who wants to have a minimal build system that I can review/understand as much as possible. Also, I still like to build all the Docker (base/child) images myself. Let me know what you think: https://github.com/hashbang/aosp-build/pull/7

lrvick committed:
config update + fix to release matching regex

ypid committed:
[scripts/shell] Preserve useful shell functions from build/envsetup.sh

ypid committed:
[scripts/build] Reorder patching more logically (echo and patching)

ypid committed:
[scripts/config] Order Python imports

ypid committed:
[scripts/config] Update config['datetime'] on every `config` run

Closes: https://github.com/hashbang/os/issues/20

ypid committed:
[scripts/build] Only copy chromium overlay if there is a overlay

ypid committed:
[scripts/release] Drop Pixel 1 support

ypid committed:
[scripts/release] Update fastboot version check factory image

Ref: sargo-QP1A.191105.003-factory-267092f1.zip

lrvick committed:
add docs on new backends

lrvick committed:
cleanup redundant backends block

lrvick committed:
bump to 10.0.0_r10

lrvick committed:
Merge branch 'machine-abstraction-rework' of https://github.com/ypid/aosp-build into ypid-machine-abstraction-rework

lrvick committed:
Merge branch 'ypid-machine-abstraction-rework' into machine-abstraction

lrvick committed:
RIP Pixel 1

lrvick committed:
drop helm/k8s tools as docker-machine covers all needed backend use cases for now

lrvick committed:
make clean no longer exit one, and just clean what needs cleaned

ypid committed:
[scripts/environment] Fix faketime. Time travel from year 0019 to 2019

ypid committed:
[scripts/release] With fastbootd reboot is required to lock bootloader

ypid committed:
[scripts/release] AVB custom key needs to be flashed

Revert "[scripts/release] Drop Pixel 1 support"

This reverts commit e1bdece1f18fdb764bcbb0bc37dc108de8c59782.

ypid committed:
[scripts/release] Fix ShellCheck warning

ypid committed:
[scripts/fetch] Remove deprecated --force-broken option from repo sync

warning: -f/--force-broken is now the default behavior, and the options are deprecated

ypid committed:
[scripts/retry] Use func by sj26 which exits with cmd rc when all fail

The issue was that repo sync failed in my case so I build images which
where not in alliance with base.xml.

ypid committed:
[Makefile] `repo diff` Use expanded switch `--absolute` instead of `-u`

ypid committed:
[scripts/build] Reorder functions to match call order

ypid commented:
cc: @lsfxz you might also be interested in the work that Hashbang OS is doing (ref: https://github.com/lsfxz/AOSP-Build-Environment-Docker).

I also informed @Trumeet in https://github.com/Trumeet/AOSP-Build-Environment-Docker/issues/3.

ypid committed:
Give help how to generate the packages.list

ypid committed:
[scripts/build] Include timestamps in logs by enabling Bash tracing

Example log line:

```
+ 2019-11-17 17:49:34+00:00 make -j 2 target-files-package brillo_update_payload
```

lrvick committed:
Merge branch 'ypid-machine-abstraction-rework2' into machine-abstraction

lrvick commented:
@drGrove Considering the Android 9 setup is no longer updated and has known CVEs I think we should just do a hard cut to this into master at this point given that it has been tested by myself and ypid.

We keep merging PRs into this branch and it might as well be master at this point.

Thoughts?

lrvick committed:
address comments by @drgrove

lrvick commented:
@drGrove addressed all your comments in last commit

lrvick committed:
bump datetime
